cli/util/uploader.go

```diff
@@ -0,0 +1,432 @@
+			defer wg.Done()
+			for blk := range blockChan {
+				err := retry(func() error {
+					bw := io.NewBufBinWriter()
+					blk.EncodeBinary(bw.BinWriter)
+					if bw.Err != nil {
+						return fmt.Errorf("failed to encode block: %w", bw.Err)
+					}
+
+					attrs := []object.Attribute{
+						*object.NewAttribute(attr, strconv.Itoa(int(blk.Index))),
+						*object.NewAttribute("primary", strconv.Itoa(int(blk.PrimaryIndex))),
+						*object.NewAttribute("hash", blk.Hash().Reverse().String()),
+						*object.NewAttribute("prevHash", blk.PrevHash.StringLE()),
+						*object.NewAttribute("timestamp", strconv.FormatUint(blk.Timestamp, 10)),
+					}
+					return uploadObj(ctx.Context, *p, signer, acc.PrivateKey().GetScriptHash(), containerID, bw.Bytes(), attrs)
+				})
+				if err != nil {
+					errorChan <- err
+					return
+				}
+
+				if blk.Index%1000 == 0 {
+					fmt.Fprintf(ctx.App.Writer, "[%s] Successfully uploaded block: %d\n", time.Now().Format(time.RFC3339), blk.Index)
+				}
+			}
+		}()
+	}
+
+	// Wait for all workers to finish
+	go func() {
+		wg.Wait()
+		close(doneChan)
+	}()
+
+	// Wait for completion or error
+	select {
+	case err := <-errorChan:
+		return cli.Exit(fmt.Sprintf("Upload error: %v", err), 1)
+	case <-doneChan:
+		err = updateIndexFiles(ctx, *p, containerID, *acc, signer, uint(currentBlockHeight), indexFileAttribute, attr)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to update index files after upload: %v", err), 1)
+		}
+		fmt.Fprintln(ctx.App.Writer, "Upload completed successfully.")
+		return nil
+	}
+}
+
+type searchResult struct {
+	startIndex int
+	numOIDs    int
+	err        error
+}
+
+// fetchMaxBlockIndex searches the container for the maximum block index.
+func fetchMaxBlockIndex(ctx context.Context, p *pool.Pool, containerID cid.ID, priv *keys.PrivateKey, attributeKey string) (int, error) {
+	var wg sync.WaitGroup
+	height := 0
+	for {
+		results := make([]searchResult, maxParallelSearches)
+		for i := range maxParallelSearches {
+			startIndex := height + i*searchBatchSize
+			endIndex := startIndex + searchBatchSize - 1
+
+			wg.Add(1)
+			go func(i, startIndex, endIndex int) {
+				defer wg.Done()
+
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(attributeKey, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+				filters.AddFilter(attributeKey, fmt.Sprintf("%d", endIndex), object.MatchNumLE)
+				prm.SetFilters(filters)
+
+				objectIDs, err := neofs.ObjectSearch(ctx, p, priv, containerID.String(), prm)
+				results[i] = searchResult{startIndex: startIndex, numOIDs: len(objectIDs), err: err}
+			}(i, startIndex, endIndex)
+		}
+
+		wg.Wait()
+
+		for _, res := range results {
+			if res.err != nil {
+				return 0, res.err
+			}
+			if res.numOIDs < searchBatchSize {
+				if res.startIndex == 0 && res.numOIDs == 0 {
+					return -1, nil
+				}
+				// Return the start index of the first incomplete batch
+				return res.startIndex, nil
```

Comment:
It's because of :
`		for blockIndex := maxBlockIndex + 1 + i; blockIndex <= int(currentBlockHeight); blockIndex += numProducers {`
now we can remove this +1 and harmlessly reupload one block because it won't differ from the existing one.